File: src/backend/models/schemas.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime

# Chest schemas
class ChestBase(BaseModel):
    name: str

class ChestCreate(ChestBase):
    pass

class ChestUpdate(BaseModel):
    name: Optional[str] = None

class Chest(ChestBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        orm_mode = True

# Source schemas
class SourceBase(BaseModel):
    name: str
    type: str  # TXT, URL, FILE
    content: Optional[str] = None
    is_enabled: Optional[bool] = True

class SourceCreate(SourceBase):
    chest_id: int

class SourceUpdate(BaseModel):
    name: Optional[str] = None
    type: Optional[str] = None
    content: Optional[str] = None
    is_enabled: Optional[bool] = None

class Source(SourceBase):
    id: int
    chest_id: int
    content_hash: Optional[str] = None
    created_at: datetime

    class Config:
        orm_mode = True

# Chat message schemas
class ChatMessageBase(BaseModel):
    role: str  # USER or ASSISTANT
    content: str
    sources_used: Optional[list] = None

class ChatMessageCreate(ChatMessageBase):
    chest_id: int

class ChatMessage(ChatMessageBase):
    id: int
    chest_id: int
    created_at: datetime

    class Config:
        orm_mode = True

# RAG query schemas
class RAGQuery(BaseModel):
    question: str
    chest_id: int

class RAGResponse(BaseModel):
    answer: str
    sources_used: list

File: src/frontend/src/lib/models/schemas.ts
// These schemas mirror the backend Pydantic models for TypeScript type safety
export interface ChestBase {
	name: string;
}

export interface ChestCreate extends ChestBase {
}

export interface ChestUpdate {
	name?: string;
}

export interface Chest extends ChestBase {
	id: number;
	created_at: string; // ISO date string
	updated_at: string | null;
}

export interface SourceBase {
	name: string;
	type: 'TXT' | 'URL' | 'FILE';
	content?: string;
	is_enabled?: boolean;
}

export interface SourceCreate extends SourceBase {
	chest_id: number;
}

export interface SourceUpdate {
	name?: string;
	type?: 'TXT' | 'URL' | 'FILE';
	content?: string;
	is_enabled?: boolean;
}

export interface Source extends SourceBase {
	id: number;
	chest_id: number;
	content_hash?: string | null;
	created_at: string;
}

export interface ChatMessageBase {
	role: 'USER' | 'ASSISTANT';
	content: string;
	sources_used?: number[] | null;
}

export interface ChatMessageCreate extends ChatMessageBase {
	chest_id: number;
}

export interface ChatMessage extends ChatMessageBase {
	id: number;
	chest_id: number;
	created_at: string;
}

export interface RAGQuery {
	question: string;
	chest_id: number;
}

export interface RAGResponse {
	answer: string;
	sources_used: number[];
}